src/modules/rover_differential/DifferentialVelControl/DifferentialVelControl.cpp

```diff
@@ -0,0 +1,235 @@
+	rover_velocity_status_s rover_velocity_status;
+	rover_velocity_status.timestamp = _timestamp;
+	rover_velocity_status.measured_speed_body_x = _vehicle_speed_body_x;
+	rover_velocity_status.adjusted_speed_body_x_setpoint = _speed_setpoint.getState();
+	rover_velocity_status.measured_speed_body_y = _vehicle_speed_body_y;
+	rover_velocity_status.adjusted_speed_body_y_setpoint = NAN;
+	rover_velocity_status.pid_throttle_body_x_integral = _pid_speed.getIntegral();
+	rover_velocity_status.pid_throttle_body_y_integral = NAN;
+	_rover_velocity_status_pub.publish(rover_velocity_status);
+}
+void DifferentialVelControl::updateSubscriptions()
+{
+	if (_vehicle_control_mode_sub.updated()) {
+		_vehicle_control_mode_sub.copy(&_vehicle_control_mode);
+	}
+
+	if (_vehicle_attitude_sub.updated()) {
+		vehicle_attitude_s vehicle_attitude{};
+		_vehicle_attitude_sub.copy(&vehicle_attitude);
+		_vehicle_attitude_quaternion = matrix::Quatf(vehicle_attitude.q);
+		_vehicle_yaw = matrix::Eulerf(_vehicle_attitude_quaternion).psi();
+	}
+
+	if (_vehicle_local_position_sub.updated()) {
+		vehicle_local_position_s vehicle_local_position{};
+		_vehicle_local_position_sub.copy(&vehicle_local_position);
+		const Vector3f velocity_in_local_frame(vehicle_local_position.vx, vehicle_local_position.vy, vehicle_local_position.vz);
+		const Vector3f velocity_in_body_frame = _vehicle_attitude_quaternion.rotateVectorInverse(velocity_in_local_frame);
+		_vehicle_speed_body_x = fabsf(velocity_in_body_frame(0)) > _param_ro_speed_th.get() ? velocity_in_body_frame(0) : 0.f;
+		_vehicle_speed_body_y = fabsf(velocity_in_body_frame(1)) > _param_ro_speed_th.get() ? velocity_in_body_frame(1) : 0.f;
+	}
+
+}
+
+void DifferentialVelControl::generateVelocitySetpoint()
+{
+	trajectory_setpoint_s trajectory_setpoint{};
+	_trajectory_setpoint_sub.copy(&trajectory_setpoint);
+
+	if (_offboard_control_mode_sub.updated()) {
+		_offboard_control_mode_sub.copy(&_offboard_control_mode);
+	}
+
+	const bool offboard_vel_control = _offboard_control_mode.velocity && !_offboard_control_mode.position;
+
+	const Vector2f velocity_in_local_frame(trajectory_setpoint.velocity[0], trajectory_setpoint.velocity[1]);
+
+	if (offboard_vel_control && velocity_in_local_frame.isAllFinite()) {
+		differential_velocity_setpoint_s differential_velocity_setpoint{};
+		differential_velocity_setpoint.timestamp = _timestamp;
+		differential_velocity_setpoint.velocity_ned[0] = velocity_in_local_frame(0);
+		differential_velocity_setpoint.velocity_ned[1] = velocity_in_local_frame(1);
+		differential_velocity_setpoint.backwards = false;
+		_differential_velocity_setpoint_pub.publish(differential_velocity_setpoint);
+	}
+}
+
+void DifferentialVelControl::generateAttitudeAndThrottleSetpoint()
+{
+	if (_differential_velocity_setpoint_sub.updated()) {
+		_differential_velocity_setpoint_sub.copy(&_differential_velocity_setpoint);
+	}
+
+	// Catch spot turning
+	if (PX4_ISFINITE(_differential_velocity_setpoint.yaw)) {
```

Comment:
In this moment, `_differential_velocity_setpoint.velocity_ned` can not be finite and is thus not handled below? Otherwise that would overwrite it no?